src/components/tab-group/tab-group.test.ts

```diff
@@ -0,0 +1,291 @@
+import { elementUpdated, expect, fixture, waitUntil } from '@open-wc/testing';
+import { html } from 'lit';
+import { clickOnElement } from '../../internal/test';
+import { queryByTestId } from '../../internal/test/data-testid-helpers';
+import type SlTabPanel from '../tab-panel/tab-panel';
+import type SlTab from '../tab/tab';
+import type SlTabGroup from './tab-group';
+import type { HTMLTemplateResult } from 'lit';
+
+interface ClientRectangles {
+  body?: DOMRect;
+  navigation?: DOMRect;
+}
+
+const wait = async (delayInMs = 0): Promise<void> => {
+  return new Promise(resolve => {
+    window.setTimeout(() => resolve(), delayInMs);
+  });
+};
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+const waitForScrollButtonsToBeRendered = (tabGroup: SlTabGroup): Promise<any> => {
+  return waitUntil(() => {
+    const scrollButtons = tabGroup.shadowRoot?.querySelectorAll('sl-icon-button');
+    return scrollButtons?.length === 2;
+  });
+};
+
+const getClientRectangles = (tabGroup: SlTabGroup): ClientRectangles => {
+  const shadowRoot = tabGroup.shadowRoot;
+  if (shadowRoot) {
+    const nav = shadowRoot.querySelector<HTMLElement>('[part=nav]');
+    const body = shadowRoot.querySelector<HTMLElement>('[part=body]');
+    return {
+      body: body?.getBoundingClientRect(),
+      navigation: nav?.getBoundingClientRect()
+    };
+  }
+  return {};
+};
+
+const waitForScrollingToEnd = (element: Element, timeoutInMs = 500): Promise<void> => {
+  let lastLeft = element.scrollLeft;
+  let lastTop = element.scrollTop;
+  let framesWithoutChange = 0;
+  return new Promise((resolve, reject) => {
+    const timeout = window.setTimeout(() => {
+      reject(new Error('Waiting for scroll end timed out'));
+    }, timeoutInMs);
+    function checkScrollingChanged() {
+      if (element.scrollLeft !== lastLeft || element.scrollTop !== lastTop) {
+        framesWithoutChange = 0;
+        lastLeft = window.scrollX;
+        lastTop = window.scrollY;
+      } else {
+        framesWithoutChange++;
+        if (framesWithoutChange >= 20) {
+          clearTimeout(timeout);
+          resolve();
+        }
+      }
+      window.requestAnimationFrame(checkScrollingChanged);
+    }
+    checkScrollingChanged();
+  });
+};
+
+const isElementVisibleFromScrolling = (outerElement: Element, innerElement: Element): boolean => {
+  const outerRect = outerElement.getBoundingClientRect();
+  const innerRect = innerElement.getBoundingClientRect();
+  return (
+    outerRect.top <= innerRect.bottom &&
+    innerRect.top <= outerRect.bottom &&
+    outerRect.left <= innerRect.right &&
+    innerRect.left <= outerRect.right
+  );
+};
```

Comment:
Same here. Is this worth splitting out or not so much?

Reply:
Extracted